engine-types/src/types.rs

```diff
@@ -1,14 +1,129 @@
-use super::{str, vec, Add, Address, String, Sub, Vec, U256};
+use crate::{str, vec, Address, AsMut, AsRef, Deref, DerefMut, Display, String, Vec, U256};
 use borsh::{BorshDeserialize, BorshSerialize};
 
+use crate::fmt::Formatter;
+pub use evm_core::Opcode;
+
 pub type Balance = u128;
 pub type RawAddress = [u8; 20];
 pub type RawU256 = [u8; 32]; // Big-endian large integer type.
 pub type RawH256 = [u8; 32]; // Unformatted binary data of fixed length.
 pub type EthAddress = [u8; 20];
-pub type Gas = u64;
 pub type StorageUsage = u64;
 
+#[derive(Default, Debug, Clone, Copy, Eq, PartialEq, Ord, PartialOrd)]
+/// Near gas type which wraps an underlying u64.
+pub struct NearGas(u64);
+
+impl Display for NearGas {
+    fn fmt(&self, f: &mut Formatter<'_>) -> crate::fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl Deref for NearGas {
+    type Target = u64;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl DerefMut for NearGas {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+impl AsRef<u64> for NearGas {
+    fn as_ref(&self) -> &u64 {
+        &self.0
+    }
+}
+
+impl AsMut<u64> for NearGas {
+    fn as_mut(&mut self) -> &mut u64 {
+        &mut self.0
+    }
+}
+
+impl From<u64> for NearGas {
```

Comment:
I don't think we should implement `From` either. It allows developers to fix some compilation errors by simply writing `.into()` without necessarily thinking about it very much. They should ask "why do I need a `NearGas` here and don't have one?" Sometimes it might be the case that they legitimately need to convert a `u64` into a `NearGas`, but then they should explicitly call `new()`.

Newtypes are about providing type safety and every "convenience" method that allows converting to/from the raw type undermines the safety it is providing.

Reply:
Thats precisely why I only added it for the underlaying type. I didn't want it to work for any other type.

Reply:
I think even having it for the underlying type is too much. We should always be forced to ask the question about why the compiler is given an error; if there is any easy escape hatch (like `.into()`) then we might be lazy and not ask it.

Again, it should always be possible to lift a raw value into the gas value because we will need to do it sometimes (namely at the boundary of our application), but this should always be made explicit in the syntax (eg `NearGas::new(number)`) because it shouldn't happen very often.